File: backend/gemini_client.go
package main

import (
	"context"
	"fmt"
	"os"

	"github.com/google/generative-ai-go/genai"
	"google.golang.org/api/option"
)

const geminiPro string = "gemini-pro"

var explicitMode = []*genai.SafetySetting{
	{Category: genai.HarmCategoryHarassment, Threshold: genai.HarmBlockNone},
	{Category: genai.HarmCategoryHateSpeech, Threshold: genai.HarmBlockNone},
	{Category: genai.HarmCategorySexuallyExplicit, Threshold: genai.HarmBlockNone},
	{Category: genai.HarmCategoryDangerousContent, Threshold: genai.HarmBlockNone},
}

type GeminiClient struct {
	*genai.ChatSession
}

func NewGeminiClient(ctx context.Context, apiKey string) *GeminiClient {
	client, err := genai.NewClient(ctx, option.WithAPIKey(apiKey))
	if err != nil {
		panic(err)
	}

	model := client.GenerativeModel(geminiPro)
	model.SafetySettings = explicitMode
	cs := model.StartChat()
	return &GeminiClient{
		ChatSession: cs,
	}
}

func (g *GeminiClient) SendMessageStream(ctx context.Context, msg string) *genai.GenerateContentResponseIterator {
	return g.ChatSession.SendMessageStream(ctx, genai.Text(msg))
}

func getApiKey() string {
	return os.Getenv("GEMINI_AI_API_KEY")
}

func ParseGeminiResponse(resp *genai.GenerateContentResponse) string {
	var response string
	for _, cand := range resp.Candidates {
		if cand.Content != nil {
			for _, part := range cand.Content.Parts {
				response += fmt.Sprintf("%v", part)
			}
		}
	}
	return response
}

File: backend/src/services/gemini-client.ts
import { ChatSession, Content, GenerativeModel, GoogleGenerativeAI, HarmBlockThreshold, HarmCategory, SafetySetting } from "@google/generative-ai";

const safetySettings: Array<SafetySetting> = [
    {
        category: HarmCategory.HARM_CATEGORY_HARASSMENT,
        threshold: HarmBlockThreshold.BLOCK_NONE
    },
    {
        category: HarmCategory.HARM_CATEGORY_HATE_SPEECH,
        threshold: HarmBlockThreshold.BLOCK_NONE
    },
    {
        category: HarmCategory.HARM_CATEGORY_SEXUALLY_EXPLICIT,
        threshold: HarmBlockThreshold.BLOCK_NONE
    },
    {
        category: HarmCategory.HARM_CATEGORY_DANGEROUS_CONTENT,
        threshold: HarmBlockThreshold.BLOCK_NONE
    }
];

class GeminiClient {
    private model!: GenerativeModel;
    private chatSession!: ChatSession;

    constructor(apiKey: string) {
        const genai = new GoogleGenerativeAI(apiKey);
        this.model = genai.getGenerativeModel({ model: 'gemini-pro' });
    }

    startChat(history: any = []): GeminiClient {
        this.chatSession = this.model.startChat({
            history: history,
            safetySettings
        });
        return this;
    }

    async sendMessage(message: string) {
        const result = await this.chatSession?.sendMessage(message);
        return result.response.text();
    }

    getHistory(): Promise<Array<Content>> {
        return this.chatSession.getHistory();
    }
}

export default GeminiClient;